logicalplan/set_projection_labels.go

```diff
@@ -0,0 +1,152 @@
+package logicalplan
+
+import (
+	"maps"
+	"slices"
+
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/util/annotations"
+	"github.com/thanos-io/promql-engine/query"
+)
+
+// setProjectionLabels is an optimizer that sets the projection labels for all vector selectors.
+// If a projection is already set as a matcher, it will be materialized in the selector
+// and the matcher will be removed.
+// This is useful for sending projections as part of a remote query.
+type setProjectionLabels struct{}
+
+const seriesIDLabel = "__series__id"
+
+func (s setProjectionLabels) Optimize(expr Node, _ *query.Options) (Node, annotations.Annotations) {
+	var hasProjections bool
+	TraverseBottomUp(nil, &expr, func(_ *Node, current *Node) bool {
+		switch e := (*current).(type) {
+		case *VectorSelector:
+			// Check if a projection is already set in the node.
+			hasProjections = e.Projection.Include || len(e.Projection.Labels) > 0
+		}
+		return hasProjections
+	})
+	if hasProjections {
+		return expr, annotations.Annotations{}
+	}
+
+	var projection Projection
+	return s.optimize(expr, projection)
+}
+
+func (s setProjectionLabels) optimize(expr Node, projection Projection) (Node, annotations.Annotations) {
+	var stop bool
+	Traverse(&expr, func(current *Node) {
+		if stop {
+			return
+		}
+		switch e := (*current).(type) {
+		case *Aggregation:
+			switch e.Op {
+			case parser.TOPK, parser.BOTTOMK:
+				projection = Projection{}
+			default:
+				projection = Projection{
+					Labels:  slices.Clone(e.Grouping),
+					Include: !e.Without,
+				}
```

Comment:
May I know how your scanner is implemented when it is a without? Does it return series hash label in this case?

https://github.com/thanos-io/promql-engine/pull/549/files#diff-2d378c97867270a9cf6883930a4a229b3928d637e4c5c558abf5f4bd913f2f8cR41

This is what I implemented as a mock storage which does projection and always return series hash when projection is available. So maybe there is some difference here.

Reply:
When it is `without`, it returns the hash over all labels, and the labels which are not in the `Labels` slice.

Reply:
Do you need to include series hash in the grouping label for without? So that it can be removed after the aggregation.

Or you have another place to remove series hash label?

Reply:
I'll need to take a look at the details, will post a bit later.

Reply:
I believe we dont need to include because the hash will be just an implementation detail of the storage. We use it for things like: partitioning, map series to iterators, etc.

Reply:
I think it is required as it affects the query correctness.

I am trying to use the optimizer with the query fuzzer but it seems it just fail on without queries if hash label is not excluded here.

If it requires certain behavior from the storage to remove the hash probably need to document it clearly to make sure there is no correctness problem.

Reply:
We include the hash label only when Include is true. We also have an explicit select for the series hash label for some binary joins as shown in tests.

Reply:
If we don't have hash label when without then how do we pass the duplicate labels check in the engine? Do we have to disable it.

I am trying to run some correctness tests and see some issues based on my setup. Maybe you can try to add a similar correctness test using a mock storage and see if there is any issue.

Reply:
I see. But I can still see some correctness issues after disabling duplicate check so maybe you can take a closer look

Reply:
Let's disable them here too - with a hash column we won't get issues though since we never drop it i think